blscurve/bls_batch_verifier.nim

```diff
@@ -31,116 +31,60 @@ import
-       batcher: var BatchedBLSVerifierCache,
-       public_key: PublicKey,
-       message: array[32, byte],
-       signature: Signature
-     ) {.inline.} =
-  ## Include a (public key, message, signature) triplet
-  ## to the batch for verification.
-  ##
-  ## Assumes subgroup checks
-  ## and infinity checks were done prior.
-  ## This is currently guaranteed
-  ## on deserialization of public keys and signatures.
-  ##
-  ## The proof-of-possession MUST be verified before calling this function.
-  ## For Eth2, the public key must correspond to a valid deposit.
-  batcher.sets.add SignatureSet(
-    signature: signature,
-    pubkey: public_key,
-    message: message
-  )
-
-func add*(
-       batcher: var BatchedBLSVerifierCache,
-       public_keys: openarray[PublicKey],
-       message: array[32, byte],
-       signature: Signature
-     ): bool =
-  ## Include a (array of public keys, message, signature) triplet
-  ## to the batch for verification.
-  ##
-  ## All public keys sign the same message
-  ## and signature is their aggregated signature
-  ##
-  ## Returns false if no public keys are passed
-  ## Returns true otherwise
-  ##
-  ## Assumes subgroup checks
-  ## and infinity checks were done prior.
-  ## This is currently guaranteed
-  ## on deserialization of public keys and signatures.
-  ##
-  ## The proof-of-possession MUST be verified before calling this function.
-  ## For Eth2, the public key must correspond to a valid deposit.
-  if publicKeys.len == 0:
-    return false
-
-  var aggAffine{.noInit.}: PublicKey
-  if not aggAffine.aggregateAll(publicKeys):
-    return false
-
-  batcher.sets.add SignatureSet(
-    signature: signature,
-    pubkey: aggAffine,
-    message: message
-  )
-
-  return true
-
 # Serial Batch Verifier
 # ----------------------------------------------------------------------
 
 func batchVerifySerial*(
-       batcher: var BatchedBLSVerifierCache,
+       cache: var BatchedBLSVerifierCache,
+       input: openArray[SignatureSet],
        secureRandomBytes: array[32, byte]
      ): bool =
   ## Single-threaded batch verification
-  if batcher.sets.len == 0:
+  ## This will verify all the inputs (PublicKey, message, Signature) triplets
+  ##  at once and return true if verification is successful.
+  ## If unsuccessful:
+  ## - The input was empty
+  ## - One or more of the inputs was invalid on aggregation
+  ## - One or more of the inputs had an invalid signature
+  ## If knowing which input was problematic is required, they must be checked one by one.
+  if input.len == 0:
     # Spec precondition
     return false
 
-  batcher.batchContexts.setLen(1)
-  template ctx: untyped = batcher.batchContexts[0]
-  batcher.batchContexts[0].init(secureRandomBytes, "")
+  cache.batchContexts.setLen(1)
+  template ctx: untyped = cache.batchContexts[0]
+  cache.batchContexts[0].init(secureRandomBytes, "")
 
   # Accumulate line functions
-  for i in 0 ..< batcher.sets.len:
+  for i in 0 ..< input.len:
```

Comment:
why not `for i in input`?

Reply:
I don't want to put 96 (PublicKey) + 16 (seq[byte]) + 192 (Signature) bytes on the stack.
https://github.com/nim-lang/Nim/issues/14421